# Explore 14k

This is a first systematic exploration of a reasonably sized reasonably diverse set of data for interesting structure. 

1. The sample is the 14,098 data generated by the 54 codas in Generate.codas generated to depth 3 with Generate.Gen. 

2. The structure we are searching for are at least kernels, idempotents, spaces, distributive data, i.e. types and morphisms. 



In [1]:
from base import *
import Generate,Set
#
#    Starting samples that will be used 
#
codas = Generate.codas
SMALL = Generate.Gen(codas,3).set() # 14K sample 
atoms10  = Generate.OddAtoms(5)+Generate.EvenAtoms(5)
atoms100 = Generate.OddAtoms(50)+Generate.EvenAtoms(50)

In [2]:
BIG = Generate.Gen(codas,4).set()

In [3]:
print(BIG)
for s in BIG.sample(10): print(s)

1047810
wrap equiv {B 0}
post {B first} apby
aps get1 {B null}
B apif {post B}
first rep {wrap B}
{B nth1} collect pre
collect {B ap2} nor
not bin =
ap aps xor
{and B} post startswith


In [4]:
#
#    As a quick, direct test, let's look for types in SMALL 
#
types1 = SMALL.distributive(atoms10,atoms10)

In [5]:
types1e = types1.multieval(Set.eval100) 

In [6]:
types1e.empty()

2647

In [7]:
types2 = types1e.empty()

In [8]:
types3 = types2.distributive(atoms10+SMALL.sample(10),atoms10+SMALL.sample(10))

In [9]:
types3

2647

In [10]:
types3e = types3.multieval(Set.eval100)

In [11]:
types3e.empty()

2484

In [12]:
types4 = types3e.empty()

In [13]:
for s in types4.sample(10): print(s)

ap2 a
ap equiv
nth1 {B counts}
startswith {B app}
pass {B get1}
if if
nth1 {not B}
nand app
null {B counts}
rep {B a}


In [14]:
types5 = types4.distributive(atoms10+BIG.sample(20),atoms10+BIG.sample(20))

In [15]:
types5e = types5.multieval(Set.eval100)

In [16]:
types5e.empty()

2484

In [17]:
for s in types5e.empty().sample(10): print(s)

ap2 last
nand {B}
startswith b
{B} get1
ap {bool B}
rep {B nth1}
if {B}
rep {B get1}
rep wrap
pass nth1


There are 2484 types, apparently.  However, many of these 
may be trivial in the sense that T:X = () for any X. 
I'll eliminate these by requiring notkernel = () against a large 
sample. This guarantees that T:X is atomic on *some* X at least.

In [18]:
types6 = types5e.empty()

In [22]:
types7 = types6.notkernel(atoms100+BIG.sample(1000))

In [23]:
types7e = types7.multieval(Set.eval100)

In [24]:
types7e.empty()

0